### PR TITLE
Network Lobbies

### PR DESCRIPTION
**Pull request by SheridanR**

### CODE REVIEW

File: src/ui/MainMenu.cpp [hunk 1]
```diff
@@ -3985,25 +4166,1324 @@ namespace MainMenu {
...
+	    directConnect = false;
+	    if (LobbyHandler.getHostingType() == LobbyHandler_t::LobbyServiceType::LOBBY_STEAM) {
+#ifdef STEAMWORKS
+		    for ( int c = 0; c < MAXPLAYERS; c++ ) {
+			    if ( steamIDRemote[c] ) {
+				    cpp_Free_CSteamID(steamIDRemote[c]);
+				    steamIDRemote[c] = NULL;
+			    }
+		    }
+		    ::currentLobbyType = k_ELobbyTypePrivate;
```
Comment (WALLOFJUSTICE) [inline src/ui/MainMenu.cpp:5245]: Invalid assignment here - two different enums. I'm sure you want to CreateLobby() using k_ELobbyTypePrivate but ::currentLobbyType should be something like `LobbyType::LobbyOnline`?

Comment (SheridanR) [reply src/ui/MainMenu.cpp:5245]: ::currentLobbyType refers to the definition in steam.hpp, so this compiles fine for me

Comment (WALLOFJUSTICE) [reply src/ui/MainMenu.cpp:5245]: My b, still was using code from Friday testing out steamworks builds now

File: src/ui/MainMenu.cpp [hunk 2]
```diff
@@ -3985,25 +4166,1324 @@ namespace MainMenu {
...
+				}
+
+				continue;
+			}
+
+			// network scan
+			else if (packetId == 'SCAN') {
+			    if (directConnect) {
+			        char hostname[256] = { '\0' };
+			        (void)gethostname(hostname, sizeof(hostname));
```
Comment (WALLOFJUSTICE) [inline src/ui/MainMenu.cpp:4616]: Note I'm getting missing definition here for `gethostname`

File: src/ui/MainMenu.cpp [hunk 3]
```diff
@@ -6054,26 +7541,118 @@ namespace MainMenu {
...
+	        // set unique game key
+	        uniqueGameKey = prng_get_uint();
+	        if (!uniqueGameKey) {
+		        uniqueGameKey++;
+	        }
+
+	        // send start signal to each player
+	        if (multiplayer == SERVER) {
+	            for (int c = 1; c < MAXPLAYERS; c++) {
+		            if (client_disconnected[c] || players[c]->isLocalPlayer()) {
```
Comment (WALLOFJUSTICE) [inline src/ui/MainMenu.cpp:7642]: Can't use `players[c]->isLocalPlayer()` here as `intro` is true and every client returns true on the lobby screen

Comment (SheridanR) [reply src/ui/MainMenu.cpp:7642]: fair enough, will just change it to `c == clientnum`

File: src/ui/MainMenu.cpp (hunk 2 above)
Comment (SheridanR) [reply src/ui/MainMenu.cpp:4616]: ah yes, forgot this is a POSIX function. Fun times.

**Pull request merged**